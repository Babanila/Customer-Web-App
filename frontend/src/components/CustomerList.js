import React from 'react'
import { useHistory } from 'react-router-dom'
import { cx, css } from 'emotion'
import SingleCustomer from './SingleCustomer'
import SingleButton from './SingleButton'
import ErrorComponent from './ErrorComponent'
import LoadingComponent from './LoadingComponent'
import { fetcher, baseUrl } from './Utils'

const customerListDiv = css`
  margin-top: 6em;
  width: 100%;
  display: flex;
  justify-content: center;
  z-index: 10;
`

const tableDiv = css`
  margin-top: 2em;
  width: 100%;
  max-width: 800px;
  display: table;
`

const tableTitleDiv = css`
  display: table-caption;
  font-size: 26px;
  font-weight: bold;
  padding-left: 10px;
  padding-bottom: 10px;
`

const titleWithBtnDiv = css`
  display: flex;
  justify-content: space-between;
  align-items: center;
`

const tableHeaderDiv = css`
  display: table-cell;
  font-size: 20px;
  font-weight: bold;
  padding: 10px;
  text-align: center;
  border-bottom: 1px solid black;
  background-color: #f6f6f6;
  @media (max-width: 420px) {
    font-size: 16px;
  }
`

const tableBodyDiv = css`
  display: table-row-group;
`

const tableNullBody = css`
  display: table-caption;
  font-size: 20px;
  color: #f54b3a;
`

const addCustomerBtnStyle = css`
  padding: 5px;
  font-size: 0.7em;
  color: #ffffff;
  border: 2px solid #009900;
  background-color: #009900;
  border-radius: 5px;
  outline: none;
  &:hover {
    cursor: pointer;
    color: #000000;
    background-color: #ffffff;
    border: 2px solid #f54b3a;
  }
`

function CustomerList() {
  const [allCustomerData, setAllCustomerData] = React.useState([])
  const [loading, setLoading] = React.useState(true)
  const [error, setError] = React.useState('')

  React.useEffect(() => {
    fetcher(baseUrl, setAllCustomerData, setError, setLoading)
  }, [baseUrl])

  const history = useHistory()
  const showCustomerDetails = (id) => history.push(`/customer/${id}`)
  const handleNewCustomer = () => history.push(`/new_customer`)

  if (error === 'Network Error') return <ErrorComponent message="Please contact your provider." />
  if (loading) return <LoadingComponent />

  return (
    <div className={cx(customerListDiv)}>
      <div className={cx(tableDiv)}>
        <div className={cx(tableTitleDiv)}>
          <div className={cx(titleWithBtnDiv)}>
            <div>Customer List</div>
            <SingleButton
              btnName="Add Customer"
              btnClick={handleNewCustomer}
              btnStyles={addCustomerBtnStyle}
            />
          </div>
        </div>
        <div className={cx(tableHeaderDiv)}>ID</div>
        <div className={cx(tableHeaderDiv)}>Firstname</div>
        <div className={cx(tableHeaderDiv)}>Lastname</div>
        <div className={cx(tableHeaderDiv)}>Last Contact</div>
        <div className={cx(tableHeaderDiv)}>Lifetime Value</div>
        {allCustomerData.length === 0 ? (
          <div className={cx(tableNullBody)}> No Customer Data in the Database</div>
        ) : (
          <div className={cx(tableBodyDiv)}>
            {allCustomerData.map((item, i) => (
              <SingleCustomer key={i} customerData={item} onClick={showCustomerDetails} />
            ))}
          </div>
        )}
      </div>
    </div>
  )
}

export default CustomerList
