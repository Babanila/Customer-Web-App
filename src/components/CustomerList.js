import React from 'react';
import CustomerDetails from './CustomerDetails';
import { cx, css } from 'emotion';

const tableDiv = css`
  width: 100%;
  display: table;
`;

const tableTitleDiv = css`
  display: table-caption;
  text-align: center;
  font-size: 24px;
  font-weight: bold;
`;

const tableHeaderDiv = css`
  display: table-cell;
  font-size: 20px;
  font-weight: bold;
  padding: 10px;
  text-align: justify;
  border-bottom: 1px solid black;
  background-color: #f6f6f6;
`;

const tableBodyDiv = css`
  display: table-row-group;
`;

function CustomerList({ allCustomerData }) {
  return (
    <div>
      <div className={cx(tableDiv)}>
        <div className={cx(tableTitleDiv)}> Customer List </div>
        <div className={cx(tableHeaderDiv)}>ID</div>
        <div className={cx(tableHeaderDiv)}>Firstname</div>
        <div className={cx(tableHeaderDiv)}>Lastname</div>
        <div className={cx(tableHeaderDiv)}>Last Contact</div>
        <div className={cx(tableHeaderDiv)}>Lifetime Value</div>
        <div className={cx(tableBodyDiv)}>
          {allCustomerData.map((item, i) => (
            <CustomerDetails key={i} customerData={item} />
          ))}
        </div>
      </div>
    </div>
  );
}

export default CustomerList;
