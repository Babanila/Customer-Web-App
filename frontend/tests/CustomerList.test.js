import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import CustomerList from '../src/components/CustomerList'

describe('Check CustomerList Component', () => {
  let sampleData, CustomerListWrapper
  let actualDiv, actualDiv1, actualDiv2

  beforeEach(() => {
    sampleData = [
      {
        customerID: 4,
        name: {
          first: 'Roxy',
          last: 'Fox'
        },
        birthday: '1979-06-30',
        gender: 'w',
        lastContact: '2017-01-29T21:08:50.700Z',
        customerLifetimeValue: 213.12
      },
      {
        customerID: 7,
        name: {
          first: 'Sarah',
          last: 'Conol'
        },
        birthday: '1986-03-21',
        gender: 'w',
        lastContact: '2017-06-01T23:28:56.782Z',
        customerLifetimeValue: 2191
      }
    ]

    CustomerListWrapper = mount(
      <MemoryRouter initialentries="{['/']}">
        <CustomerList />
      </MemoryRouter>
    )

    actualDiv = CustomerListWrapper.find('MemoryRouter')
    actualDiv1 = actualDiv.find('Router')
    actualDiv2 = actualDiv1.find('CustomerList')
  })

  it('should renders component without crashing', () => {
    const tree = renderer.create(<CustomerList />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render CustomerList component without data', () => {
    const actualDiv3 = actualDiv2.find('LoadingComponent')

    expect(CustomerList).toBeDefined()
    expect(actualDiv.length).toEqual(1)
    expect(actualDiv1.length).toEqual(1)
    expect(actualDiv2.length).toEqual(1)
    expect(actualDiv3.length).toEqual(1)
  })

  it('should render CustomerList component data', () => {
    const setSelectedCustomer = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setSelectedCustomer([...sampleData]))

    const setLoading = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setLoading(false))

    const CustomerListWrap = mount(
      <MemoryRouter initialentries="{['/']}">
        <CustomerList />
      </MemoryRouter>
    )

    const actual1 = CustomerListWrap.find('div')
    const actual2 = CustomerListWrap.find('SingleCustomer')
    const actual3 = actual2.children()
    const childDetails = actual3.at(0).simulate('click')

    expect(actual1.length).toEqual(23)
    expect(actual2.length).toEqual(2)
    expect(actual3.length).toEqual(2)
    expect(actual3.at(0).text()).toContain('4RoxyFoxSun, 29 Jan 2017 21:08:50 GMT213.12')
    expect(actual3.at(1).text()).toContain('7SarahConolThu, 01 Jun 2017 23:28:56 GMT2191')
    expect(actual3.at(0).childAt(0).text()).toEqual('4')
    expect(actual3.at(0).childAt(2).text()).toEqual('Fox')
    expect(childDetails.length).toEqual(1)
  })
})
