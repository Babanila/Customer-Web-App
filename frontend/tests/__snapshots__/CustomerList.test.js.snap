// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check CustomerList Component should renders component without crashing 1`] = `
<h3
  className="css-d8gwku"
>
  Loading ...
</h3>
`;
